### ISSUE TITLE
Inconsistent lock state, possible deadlock at Tempesta start.

### ISSUE THREAD

**Issue by keshonok** [natsys/tempesta#262]
Starting Tempesta in a VM with the following configuration with two servers, where 192.168.10.1 is the address of the host system.
```
cache off;
server 192.168.10.1:80;
server 192.168.10.1:8080;
```

Often I get the following trace:
```
Starting nginx: nginx.
Loading Tempesta kernel modules...
[    8.561053] [tdb] Start Tempesta DB
[    8.740770] [tempesta] Initializing Tempesta FW kernel module...
[    8.741413] [tempesta] init: cfg_if
[    8.741784] [tempesta] init: tls
[    8.742124] [tempesta] init: http
[    8.742486] [tempesta] init: http_sticky
[    8.745371] [tempesta] init: server
[    8.745778] [tempesta] init: client
[    8.746157] [tempesta] init: sock_srv
[    8.746493] [tempesta] init: sock_clnt
[    8.746757] [tempesta]   register cfg: filter
[    8.747070] [tempesta]   register cfg: cache
[    8.747359] [tempesta]   register cfg: http_sticky
[    8.747684] [tempesta]   register cfg: sock_srv
[    8.747995] [tempesta]   register cfg: sock_clnt
[    8.762495] [tempesta] Registering new scheduler: dummy
[    8.781291] [tempesta]   tfw_sched_hash: init
[    8.781777] [tempesta] Registering new scheduler: hash
[    8.797326] [tempesta]   tfw_sched_http: init
[    8.797646] [tempesta]   register cfg: tfw_sched_http
[    8.797983] [tempesta] Registering new scheduler: http
[    8.813431] [tempesta]   tfw_sched_rr: init
[    8.813790] [tempesta] Registering new scheduler: round-robin
Starting Tempesta...
[    8.818064] [tempesta] got state via sysctl: start
[    8.818563] [tempesta]   reading configuration file...
[    8.819071] [tempesta]   reading file: /home/natsys-lab/tfw-run/tempesta/etc/tempesta_fw.conf
[    8.820495] [tempesta]   file size: 5807 bytes
[    8.820937] [tempesta]   read by offset: 0
[    8.821390] [tempesta]   read by offset: 4096
[    8.821901] [tempesta]   read by offset: 5807
[    8.822408] [tempesta] starting all modules...
[    8.822850] [tempesta]   parsing configuration and pushing it to modules...
[    8.823725] [tempesta]   spec handle: 'cache'
[    8.824403] [tempesta]   spec handle: 'server'
[    8.825092] [tempesta]   new server group: 'default'
[    8.825782] [tempesta]   spec handle: 'server'
[    8.826229] [tempesta]   use default entry: 'filter_tbl_size 16777216;'
[    8.826873] [tempesta]   spec handle: 'filter_tbl_size'
[    8.827363] [tempesta]   use default entry: 'filter_db /opt/tempesta/db/filter.tdb;'
[    8.828108] [tempesta]   spec handle: 'filter_db'
[    8.828572] [tempesta]   use default entry: 'cache_size 268435456;'
[    8.829157] [tempesta]   spec handle: 'cache_size'
[    8.829612] [tempesta]   use default entry: 'cache_db /opt/tempesta/db/cache.tdb;'
[    8.830317] [tempesta]   spec handle: 'cache_db'
[    8.830762] [tempesta]   use default entry: 'listen 80;'
[    8.831265] [tempesta]   spec handle: 'listen'
[    8.831692] [tempesta]   starting modules...
[    8.832151] [tempesta]   mod_start(): filter
[    8.839426] [tdb] Opened table /opt/tempesta/db/filter.tdb: size=16777216 rec_size=20 base=ffff880033000000
[    8.840143] [tempesta]   mod_start(): cache
[    8.840429] [tempesta]   mod_start(): http_sticky
[    8.840752] [tempesta]   mod_start(): sock_srv
[    8.841132] [tempesta] ERROR: can't initiate a connect to server: error -101
[    8.841639] [sync_sockets]   Close socket ffff880038c9c940 (account=1)
[    8.842098] [sync_sockets]   __ss_do_close: sk ffff880038c9c940, sk->sk_socket           (null), state (Close)
[    8.842806] [tempesta] ERROR: can't initiate a connect to server: error -101
[    8.843300] [sync_sockets]   Close socket ffff880038c9c940 (account=1)
[    8.843745] [sync_sockets]   __ss_do_close: sk ffff880038c9c940, sk->sk_socket           (null), state (Close)
[    8.844424] [tempesta] ERROR: can't initiate a connect to server: error -101
[    8.844898] [sync_sockets]   Close socket ffff880038c9c940 (account=1)
[    8.845348] [sync_sockets]   __ss_do_close: sk ffff880038c9c940, sk->sk_socket           (null), state (Close)
[    8.846039] [tempesta] ERROR: can't initiate a connect to server: error -101
[    8.846518] [sync_sockets]   Close socket ffff880038c9c940 (account=1)
[    8.846957] [sync_sockets]   __ss_do_close: sk ffff880038c9c940, sk->sk_socket           (null), state (Close)
[    8.847628] [tempesta] ERROR: can't initiate a connect to server: error -101
[    8.848113] [sync_sockets]   Close socket ffff880038c9c940 (account=1)
[    8.848551] [sync_sockets]   __ss_do_close: sk ffff880038c9c940, sk->sk_socket           (null), state (Close)
[    8.849233] [tempesta] ERROR: can't initiate a connect to server: error -101
[    8.849707] [sync_sockets]   Close socket ffff880038c9c940 (account=1)
[    8.850151] [sync_sockets]   __ss_do_close: sk ffff880038c9c940, sk->sk_socket           (null), state (Close)
[    8.850835] [tempesta] ERROR: can't initiate a connect to server: error -101
[    8.851311] [sync_sockets]   Close socket ffff880038c9c940 (account=1)
[    8.851753] [sync_sockets]   __ss_do_close: sk ffff880038c9c940, sk->sk_socket           (null), state (Close)
[    8.852427] [tempesta] ERROR: can't initiate a connect to server: error -101
[    8.852894] [sync_sockets]   Close socket ffff880038c9c940 (account=1)
[    8.853325] [sync_sockets]   __ss_do_close: sk ffff880038c9c940, sk->sk_socket           (null), state (Close)
[    8.853974] [tempesta]   mod_start(): sock_clnt
[    8.854271] [tempesta] Open listen socket on: 0.0.0.0
[    8.854622] [tempesta]   start listening on socket: sk=ffff880038c9c940
[    8.855063] [tempesta]   mod_start(): tfw_sched_http
[    8.855391] [tempesta]   spec handle: 'sched_http_rules'
[    8.855738] [tempesta]   tfw_sched_http: begin sched_http_rules
[    8.856135] [tempesta]   spec handle: 'match'
[    8.856423] [tempesta]   tfw_sched_http: parsed rule: match  'default'=ffff880038bc36c0  '*'=1  '*'=1  '*'
[    8.857044] [tempesta]   tfw_sched_http: finish sched_http_rules
[    8.857434] [tempesta] modules are started
net.tempesta.state = start
done
[    9.840449] [tempesta] ERROR: can't initiate a connect to server: error -101
[    9.842769] 
[    9.843276] =================================
[    9.844046] [ INFO: inconsistent lock state ]
[    9.844046] 3.10.10-tempesta+ #31 Tainted: G           O
[    9.844046] ---------------------------------
[    9.844046] inconsistent {SOFTIRQ-ON-W} -> {IN-SOFTIRQ-W} usage.
[    9.844046] swapper/0/0 [HC0[0]:SC1[1]:HE1:SE0] takes:
[    9.844046]  (slock-AF_INET/1){+.?...}, at: [<ffffffffa0036799>] ss_close+0x44/0x6e [tempesta_fw]
[    9.844046] {SOFTIRQ-ON-W} state was registered at:
[    9.844046]   [<ffffffff81074f87>] __lock_acquire+0x6c7/0x1240
[    9.844046]   [<ffffffff81076328>] lock_acquire+0x98/0xd0
[    9.844046]   [<ffffffff813c9c9a>] _raw_spin_lock_nested+0x3a/0x50
[    9.844046]   [<ffffffffa0036799>] ss_close+0x44/0x6e [tempesta_fw]
[    9.844046]   [<ffffffffa00387f9>] tfw_sock_srv_connect_try+0xea/0xf6 [tempesta_fw]
[    9.844046]   [<ffffffffa0038cbb>] tfw_sock_srv_connect_srv+0x31/0x71 [tempesta_fw]
[    9.844046]   [<ffffffffa0034b6f>] tfw_sg_for_each_srv+0x6e/0xfa [tempesta_fw]
[    9.844046]   [<ffffffffa0038d76>] tfw_sock_srv_connect_all+0x10/0x12 [tempesta_fw]
[    9.844046]   [<ffffffffa00106b2>] mod_start+0x6e/0x9b [tempesta_fw]
[    9.844046]   [<ffffffffa0010b49>] tfw_cfg_start_mods+0xf0/0x1ea [tempesta_fw]
[    9.844046]   [<ffffffffa0011046>] handle_state_change+0xfd/0x19d [tempesta_fw]
[    9.844046]   [<ffffffffa0011235>] handle_sysctl_state_io+0x14f/0x1ba [tempesta_fw]
[    9.844046]   [<ffffffff811408bf>] proc_sys_call_handler+0x8f/0xc0
[    9.844046]   [<ffffffff811408ff>] proc_sys_write+0xf/0x20
[    9.844046]   [<ffffffff810e91b3>] vfs_write+0xd3/0x1b0
[    9.844046]   [<ffffffff810e9787>] SyS_write+0x47/0x90
[    9.844046]   [<ffffffff813cb2e9>] system_call_fastpath+0x16/0x1b
[    9.844046] irq event stamp: 1696130
[    9.844046] hardirqs last  enabled at (1696130): [<ffffffff81032d05>] vprintk_emit+0x4d5/0x5b0
[    9.844046] hardirqs last disabled at (1696129): [<ffffffff8103292c>] vprintk_emit+0xfc/0x5b0
[    9.844046] softirqs last  enabled at (1696092): [<ffffffff81038409>] irq_enter+0x59/0x80
[    9.844046] softirqs last disabled at (1696093): [<ffffffff81038499>] irq_exit+0x69/0xc0
[    9.844046] 
[    9.844046] other info that might help us debug this:
[    9.844046]  Possible unsafe locking scenario:
[    9.844046] 
[    9.844046]        CPU0
[    9.844046]        ----
[    9.844046]   lock(slock-AF_INET/1);
[    9.844046]   <Interrupt>
[    9.844046]     lock(slock-AF_INET/1);
[    9.844046] 
[    9.844046]  *** DEADLOCK ***
[    9.844046] 
[    9.844046] 1 lock held by swapper/0/0:
[    9.844046]  #0:  (((&srv_conn->retry_timer))){+.-...}, at: [<ffffffff8103dc30>] call_timer_fn+0x0/0xe0
[    9.844046] 
[    9.844046] stack backtrace:
[    9.844046] CPU: 0 PID: 0 Comm: swapper/0 Tainted: G           O 3.10.10-tempesta+ #31
[    9.844046]  ffffffff81acbdf0 ffff88003fc03c88 ffffffff813c5f2d ffff88003fc03cd8
[    9.844046]  ffffffff813c368a 0000000000000001 ffffffff00000001 ffff880000000000
[    9.844046]  0000000000000006 ffffffff81565440 ffffffff81072dd0 0000000000000004
[    9.844046] Call Trace:
[    9.844046]  <IRQ>  [<ffffffff813c5f2d>] dump_stack+0x19/0x1b
[    9.844046]  [<ffffffff813c368a>] print_usage_bug+0x1f7/0x208
[    9.844046]  [<ffffffff81072dd0>] ? print_irq_inversion_bug+0x220/0x220
[    9.844046]  [<ffffffff81074026>] mark_lock+0x176/0x2a0
[    9.844046]  [<ffffffff81074ee9>] __lock_acquire+0x629/0x1240
[    9.844046]  [<ffffffff81346f8d>] ? tcp_v4_init_sock+0xd/0x30
[    9.844046]  [<ffffffffa00388d4>] ? __reset_retry_timer+0x2b/0x2b [tempesta_fw]
[    9.844046]  [<ffffffff81076328>] lock_acquire+0x98/0xd0
[    9.844046]  [<ffffffffa0036799>] ? ss_close+0x44/0x6e [tempesta_fw]
[    9.844046]  [<ffffffff813c9c9a>] _raw_spin_lock_nested+0x3a/0x50
[    9.844046]  [<ffffffffa0036799>] ? ss_close+0x44/0x6e [tempesta_fw]
[    9.844046]  [<ffffffffa0036799>] ss_close+0x44/0x6e [tempesta_fw]
[    9.844046]  [<ffffffffa00387f9>] tfw_sock_srv_connect_try+0xea/0xf6 [tempesta_fw]
[    9.844046]  [<ffffffffa00388f4>] tfw_sock_srv_connect_retry_timer_cb+0x20/0x48 [tempesta_fw]
[    9.844046]  [<ffffffff8103dc9b>] call_timer_fn+0x6b/0xe0
[    9.844046]  [<ffffffff8103dc30>] ? process_timeout+0x10/0x10
[    9.844046]  [<ffffffffa00388d4>] ? __reset_retry_timer+0x2b/0x2b [tempesta_fw]
[    9.844046]  [<ffffffff8103df33>] run_timer_softirq+0x1e3/0x230
[    9.844046]  [<ffffffff81038220>] __do_softirq+0xe0/0x1d0
[    9.844046]  [<ffffffff81038499>] irq_exit+0x69/0xc0
[    9.844046]  [<ffffffff8101d85a>] smp_apic_timer_interrupt+0x8a/0xa0
[    9.844046]  [<ffffffff813cbef2>] apic_timer_interrupt+0x72/0x80
[    9.844046]  <EOI>  [<ffffffff81023626>] ? native_safe_halt+0x6/0x10
[    9.844046]  [<ffffffff8100a8dd>] default_idle+0x1d/0x30
[    9.844046]  [<ffffffff8100b26a>] arch_cpu_idle+0xa/0x10
[    9.844046]  [<ffffffff8106817c>] cpu_startup_entry+0x15c/0x200
[    9.844046]  [<ffffffff813be57d>] rest_init+0xcd/0xe0
[    9.844046]  [<ffffffff813be4b0>] ? csum_partial_copy_generic+0x170/0x170
[    9.844046]  [<ffffffff815dada5>] start_kernel+0x352/0x35d
[    9.844046]  [<ffffffff815da85c>] ? repair_env_string+0x5e/0x5e
[    9.844046]  [<ffffffff815da120>] ? early_idt_handlers+0x120/0x120
[    9.844046]  [<ffffffff815da55b>] x86_64_start_reservations+0x2a/0x2c
[    9.844046]  [<ffffffff815da64c>] x86_64_start_kernel+0xef/0xfc
[    9.881093] [sync_sockets]   Close socket ffff880038331280 (account=1)
[    9.881512] [sync_sockets]   __ss_do_close: sk ffff880038331280, sk->sk_socket           (null), state (Close)
[    9.882152] [tempesta] Warning: server connect retry failed
[    9.882517] [tempesta] ERROR: can't initiate a connect to server: error -101
[    9.882970] [sync_sockets]   Close socket ffff880038331280 (account=1)
[    9.883386] [sync_sockets]   __ss_do_close: sk ffff880038331280, sk->sk_socket           (null), state (Close)
[    9.884027] [tempesta] Warning: server connect retry failed
[    9.884390] [tempesta] ERROR: can't initiate a connect to server: error -101
[    9.884844] [sync_sockets]   Close socket ffff880038331280 (account=1)
[    9.885264] [sync_sockets]   __ss_do_close: sk ffff880038331280, sk->sk_socket           (null), state (Close)
[    9.885962] [tempesta] Warning: server connect retry failed
[    9.886363] [tempesta] ERROR: can't initiate a connect to server: error -101
[    9.886837] [sync_sockets]   Close socket ffff880038331280 (account=1)
[    9.887277] [sync_sockets]   __ss_do_close: sk ffff880038331280, sk->sk_socket           (null), state (Close)
[    9.887949] [tempesta] Warning: server connect retry failed
[    9.888335] [tempesta] ERROR: can't initiate a connect to server: error -101
[    9.888813] [sync_sockets]   Close socket ffff880038331280 (account=1)
[    9.889252] [sync_sockets]   __ss_do_close: sk ffff880038331280, sk->sk_socket           (null), state (Close)
[    9.889934] [tempesta] Warning: server connect retry failed
[    9.890321] [tempesta] ERROR: can't initiate a connect to server: error -101
[    9.890796] [sync_sockets]   Close socket ffff880038331280 (account=1)
[    9.891236] [sync_sockets]   __ss_do_close: sk ffff880038331280, sk->sk_socket           (null), state (Close)
[    9.891910] [tempesta] Warning: server connect retry failed
[    9.892289] [tempesta] ERROR: can't initiate a connect to server: error -101
[    9.892766] [sync_sockets]   Close socket ffff880038331280 (account=1)
[    9.893225] [sync_sockets]   __ss_do_close: sk ffff880038331280, sk->sk_socket           (null), state (Close)
[    9.893911] [tempesta] Warning: server connect retry failed
[    9.894294] [tempesta] ERROR: can't initiate a connect to server: error -101
[    9.894778] [sync_sockets]   Close socket ffff880038331280 (account=1)
[    9.895219] [sync_sockets]   __ss_do_close: sk ffff880038331280, sk->sk_socket           (null), state (Close)
[    9.895891] [tempesta] Warning: server connect retry failed
[   11.884220] [sync_sockets]   ss_tcp_state_change: sk ffff880038331280, sk->sk_socket           (null), state (Established)
[   11.884976] [tempesta]   connected: 192.168.10.1:8080
[   11.888103] [sync_sockets]   ss_tcp_state_change: sk ffff880038330980, sk->sk_socket           (null), state (Established)
[   11.888858] [tempesta]   connected: 192.168.10.1:8080
[   11.889212] [sync_sockets]   ss_tcp_state_change: sk ffff880038330080, sk->sk_socket           (null), state (Established)
[   11.890115] [tempesta]   connected: 192.168.10.1:8080
[   11.892123] [sync_sockets]   ss_tcp_state_change: sk ffff880038b932c0, sk->sk_socket           (null), state (Established)
[   11.892864] [tempesta]   connected: 192.168.10.1:8080
[   11.893220] [sync_sockets]   ss_tcp_state_change: sk ffff880038b929c0, sk->sk_socket           (null), state (Established)
[   11.893984] [tempesta]   connected: 192.168.10.1
[   11.896105] [sync_sockets]   ss_tcp_state_change: sk ffff880038b920c0, sk->sk_socket           (null), state (Established)
[   11.896851] [tempesta]   connected: 192.168.10.1
[   11.897182] [sync_sockets]   ss_tcp_state_change: sk ffff880038649300, sk->sk_socket           (null), state (Established)
[   11.897951] [tempesta]   connected: 192.168.10.1
[   11.900100] [sync_sockets]   ss_tcp_state_change: sk ffff880038648a00, sk->sk_socket           (null), state (Established)
[   11.900847] [tempesta]   connected: 192.168.10.1
```

**Comment by krizhanovsky** [comment]
There is no known installations with multiple upstream servers for now, so move the bug to 0.5.

**Comment by krizhanovsky** [comment]
Linked with #236

**Comment by keshonok** [comment]
Fixed with eb1c189.
